Example of multiple waypoints

VysotskiVadim opened the pull request:
Added example of multiple way points


https://user-images.githubusercontent.com/6190346/141496923-10343ba8-86ee-4355-b943-0f827e501e09.mp4

VysotskiVadim commented on app/src/main/res/values/styles.xml:
```diff
@@ -1,6 +1,6 @@
 <resources>
   <!-- Base application theme. -->
-  <style name="AppTheme" parent="Theme.AppCompat.Light.NoActionBar">
+  <style name="AppTheme" parent="Theme.MaterialComponents.Light.NoActionBar">
```
the Material library provides many useful components. My tab layout doesn't work with the AppCompat theme. I checked other screens - seems fine with the material theme

VysotskiVadim commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+
+    fun addRegular(point: Point) {
+        waypoints.add(Waypoint(point, WaypointType.Regular))
+    }
+
+    fun addSilent(point: Point) {
+        waypoints.add(Waypoint(point, WaypointType.Silent))
+    }
+
+    fun clear() {
+        waypoints.clear()
+    }
+
+    fun waypointsIndices(): List<Int> {
+        return waypoints.mapIndexedNotNull { index, waypoint ->
+            if (waypoint.type == WaypointsSet.WaypointType.Silent) {
+                null
+            } else index
+        }
+    }
+
+    fun waypointsNames(): List<String> = waypoints.filter { it.type != WaypointType.Silent }.map {
+        when (it.type) {
+            is WaypointType.Named -> it.type.name
+            else -> ""
+        }
+    }
+
+    fun coordinatesList(): List<Point> {
+        return waypoints.map { it.point }
+    }
+
+    private sealed class WaypointType {
+        data class Named(val name: String) : WaypointType()
+        object Regular : WaypointType()
+        object Silent : WaypointType()
+    }
+
+    private data class Waypoint(val point: Point, val type: WaypointType)
+}
```
Where do you usually put helper classes? Logic of converting waypoints to params is too complex to inline it in activity.

VysotskiVadim replied:
Okay, I'm leaving it here if there are no strong opinions 🙂

kmadsen replied:
Yeah the old examples, found the need for a [helper class too](https://github.com/mapbox/mapbox-navigation-android/blob/main/examples/src/main/java/com/mapbox/navigation/examples/core/waypoints/WaypointsController.kt). IMO it's fine to have in here.

But also, the fact we need a helper class makes me think we could design a class to handle this in core-sdk.

VysotskiVadim commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+            addedWaypoints.addRegular(originPoint)
+        }
+
+        when (val index = binding.waypointTypeView.selectedTabPosition) {
+            1 -> addedWaypoints.addNamed(destination, "Example of the named waypoint")
+            0 -> addedWaypoints.addRegular(destination)
+            2 -> addedWaypoints.addSilent(destination)
+            else -> error("Not supported type of waypoint. index $index")
+        }
```
ids and tags doesn't work for tabs. As this control isn't designed to use as a radio button I had to write this code 😃

VysotskiVadim commented on app/src/test/java/com/mapbox/navigation/examples/basics/WaypointsSetTest.kt:
```diff
@@ -0,0 +1,74 @@
+        assertEquals(listOf("", "", "test named"), names)
+    }
+
+    @Test
+    fun `map regular, silent, and named waypoints to names`() {
+        val set = WaypointsSet().apply {
+            addRegular(createPoint())
+            addSilent(createPoint())
+            addNamed(createPoint(), "test named")
+        }
+
+        val names = set.waypointsNames()
+
+        assertEquals(listOf("", "test named"), names)
+    }
```
That's actually a tricky case. Silent points should be ignored in the names section. Did I understand it correct?

RingerJK replied:
yes, that's right. Silent waypoints cannot be named

VysotskiVadim commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+        val maneuvers = maneuverApi.getManeuvers(routeProgress)
+        maneuvers.fold(
+            { error ->
+                Toast.makeText(
+                    this@MultipleWaypointsActivity,
+                    error.errorMessage,
+                    Toast.LENGTH_SHORT
+                ).show()
```
I'm constantly getting `index 1 size 0` error message. Can't find more data about in the logs. Any ideas how to troubleshoot?

VysotskiVadim replied:
Oh, my bad. it's not a native code. I can debug it 🙂

VysotskiVadim replied:
`ManeuverProcessor.updateDistanceRemainingForCurrentManeuver` throws an index out bounds exception because there is an empty array`stepToManeuvers.maneuverList` and the `currentInstructionIndex` is 1.

RingerJK replied:
please cut a ticket, it looks like data issue

VysotskiVadim replied:
https://github.com/mapbox/mapbox-navigation-android/issues/5088

korshaknn commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+     */
+    private val routesObserver = RoutesObserver { routeUpdateResult ->
+        if (routeUpdateResult.routes.isNotEmpty()) {
+            // generate route geometries asynchronously and render them
```
```suggestion
            // generate route geometries and render them
```

korshaknn commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+     * Based on whether the synthesized audio file is available, the callback plays the file
+     * or uses the fall back which is played back using the on-device Text-To-Speech engine.
+     */
+    private val speechCallback =
+        MapboxNavigationConsumer<Expected<SpeechError, SpeechValue>> { expected ->
+            expected.fold(
+                { error ->
+                    // play the instruction via fallback text-to-speech engine
+                    voiceInstructionsPlayer.play(
+                        error.fallback,
+                        voiceInstructionsPlayerCallback
+                    )
+                },
+                { value ->
+                    // play the sound file from the external generator
+                    voiceInstructionsPlayer.play(
+                        value.announcement,
+                        voiceInstructionsPlayerCallback
+                    )
+                }
+            )
+        }
```
```suggestion
    private val speechCallback =
        MapboxNavigationConsumer<Expected<SpeechError, SpeechValue>> { expected ->
            expected.fold(
                { error ->
                    // play the instruction via fallback text-to-speech engine
                    playVoiceInstruction(error.fallback)
                },
                { value ->
                    // play the sound file from the external generator
                    playVoiceInstruction(value.announcement)
                }
            )
        }
    
    private fun playVoiceInstruction(speechAnnouncement: SpeechAnnouncement) {
        voiceInstructionsPlayer.play(
            speechAnnouncement,
            voiceInstructionsPlayerCallback
        )
    }
```

korshaknn commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+        }
+
+        // initialize Mapbox Navigation
+        mapboxNavigation = if (MapboxNavigationProvider.isCreated()) {
+            MapboxNavigationProvider.retrieve()
+        } else {
+            MapboxNavigationProvider.create(
+                NavigationOptions.Builder(this.applicationContext)
+                    .accessToken(getString(R.string.mapbox_access_token))
+                    // comment out the location engine setting block to disable simulation
+                    .locationEngine(replayLocationEngine)
+                    .build()
+            )
+        }
```
```suggestion
        mapboxNavigation = MapboxNavigationProvider.create(
                NavigationOptions.Builder(this.applicationContext)
                    .accessToken(getString(R.string.mapbox_access_token))
                    // comment out the location engine setting block to disable simulation
                    .locationEngine(replayLocationEngine)
                    .build()
            )
        }
```

VysotskiVadim replied:
Why don't I need to check if `MapBoxNavigation` is already created?

Zayankovsky replied:
If you close the example and immediately start another one, you will see that `onCreate` of the next example is called before `onDestroy` of previous one. And thus you will reuse `MapboxNavigation` of the previous example and it will be destroyed a couple of moments later.

VysotskiVadim replied:
You're right. There's a delay between the `finish` and `onDestroy` in activity.

VysotskiVadim replied:
BTW, why does `TurnByTurnExperienceActivity` creates the `MapboxNavigation` in the "wrong" way. Do you recommend me to fix it as well? @Zayankovsky

Zayankovsky replied:
It definitely needs to be fixed. Whether you do it in this PR or in a separate one – up to you to decide.

RingerJK commented:
cannot choose more than 1 coordinates 

logs 
```
2021-11-08 15:15:39.662 14257-3053/com.mapbox.navigation.examples E/nav-native: getRoute("https://api.mapbox.com/directions/v5/mapbox/driving-traffic/-122.3972651,37.7851283;-122.3976166,37.7865669;-122.3962466,37.7889142?access_token=pk.eyJ1IjoicmluZ2VyamsiLCJhIjoiY2p1aWxpN2p4MWE5ZTQ0cGRod2RyMG5kYSJ9.869b3-55xeJVWd1B1UhH7A&geometries=polyline6&overview=full&steps=true&bearings=45.692791%2C45%3B&continue_straight=true&annotations=congestion_numeric%2Cmaxspeed%2Cspeed%2Cduration%2Cdistance%2Cclosure&language=en&roundabout_exits=true&voice_instructions=true&banner_instructions=true&voice_units=imperial&waypoints=0%3B1%3B2&waypoint_names=%3B%3B&enable_refresh=true") failed: RouterException No suitable edges near location

```

https://user-images.githubusercontent.com/8738058/140740487-0d728cad-2917-4761-8b06-41bcdb476ecf.mp4

korshaknn commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+        // initialize voice instructions api and the voice instruction player
+        speechApi = MapboxSpeechApi(
+            this,
+            getString(R.string.mapbox_access_token),
```
```suggestion
            getMapboxAccessTokenFromResources()
```

```
private fun getMapboxAccessTokenFromResources(): String {
                return Utils.getMapboxAccessToken(this)
            }
```

VysotskiVadim replied:
do you suggest to create the `Utils` class and put `getMapboxAccessToken` there? Do you recommend to update other examples?

VysotskiVadim replied:
replaced `getString(R.string.mapbox_access_token)` by `getMapboxAccessToken ` in the whole example project.

VysotskiVadim replied:
reverted utils because of https://github.com/mapbox/mapbox-navigation-android-examples/pull/57#issuecomment-964644166
@korshaknn are you okay with this?

korshaknn commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,556 @@
+        )
+        voiceInstructionsPlayer = MapboxVoiceInstructionsPlayer(
+            this,
+            getString(R.string.mapbox_access_token),
```
the same

VysotskiVadim commented on app/src/test/java/com/mapbox/navigation/examples/basics/WaypointsSetTest.kt:
```diff
@@ -0,0 +1,74 @@
+class WaypointsSetTest {
+    @Test
+    fun `map not silent waypoints types to waypoint indices`() {
+        val set = WaypointsSet().apply {
```
not sure if it helps. I want to be able to add different amount of different types to the set.

alexander-kulikovskii replied:
sure, up to you 
but right now it seems as one provider

VysotskiVadim commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,557 @@
+                .waypointNamesList(addedWaypoints.waypointsNames())
+                // provide the bearing for the origin of the request to ensure
+                // that the returned route faces in the direction of the current user movement
+                .bearingsList(
```
fixed

VysotskiVadim commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,557 @@
+            RouteOptions.builder()
+                .applyDefaultNavigationOptions()
+                .applyLanguageAndVoiceUnitOptions(this)
+                .coordinatesList(addedWaypoints.coordinatesList())
```
This example focuses only on multiple waypoint. Let's keep it simple and let our users learn features one by one. Anyway if we find out later that the layers and the bearing is very very important and users don't realise their importance because of misleading examples - let's make the examples more complex then.
What do you think about it?

VysotskiVadim replied:
> Silent waypoints do not become actually **silent**
> 
>  device-2021-11-09-144852.mp4

That's a very good point! Silent waypoints become silent when there is at least a one regular waypoint after it. See the [video](https://user-images.githubusercontent.com/6190346/140930194-e8dade27-f718-4314-adaf-76d99278ed46.mp4)

I'm not sure how to handle the case. I see two options.
1. We can leave example as is. It's simple and demonstrates how the SDK handles different waypoints.
2. I can add following logic: If the last waypoint is silent convert it to a regular one. This will complicate the example but will make the behaviour more obvious.

@RingerJK , what would you recommend me to do?

VysotskiVadim replied:
@RingerJK , I also don't remember any documentation about this behaviour. Could it be changed? I don't know 🤔

RingerJK replied:
I vote for the second option. We should provide reliable examples, even if customers would like to copy-paste it to the app - it has to work

RingerJK commented:
Silent waypoints do not become actually **silent**
https://user-images.githubusercontent.com/8738058/140918939-bb642c61-f61f-4554-961c-08f6e88e2c65.mp4

VysotskiVadim commented:
> Silent waypoints do not become actually **silent**
> 
>  device-2021-11-09-144852.mp4

That's a very good point! Silent waypoints become silent when there is at least a one regular waypoint after it. See the [video](https://user-images.githubusercontent.com/6190346/140930194-e8dade27-f718-4314-adaf-76d99278ed46.mp4)

I'm not sure how to handle the case. I see two options.
1. We can leave example as is. It's simple and demonstrates how the SDK handles different waypoints. I also don't remember any documentation specifying this behaviour. Maybe that could be changed in the future?
2. I can add following logic: If the last waypoint is silent convert it to a regular one. This will complicate the example but will make the behaviour more obvious.

@RingerJK , what would you recommend me to do?

RingerJK commented:
I vote for the second option. We should provide reliable examples, even if customers would like to copy-paste it to the app - it has to work

VysotskiVadim commented:
> I vote for the second option. We should provide reliable examples, even if customers would like to copy-paste it to the app - it has to work

okay. Actually I found this in the docs:
> Each item in the list must be the zero-based index of an input coordinate, and the list must include 0 (the index of the first coordinate) and the index of the last coordinate.

So the first option is just incorrect. I will fix it

kmadsen commented on app/src/main/java/com/mapbox/navigation/examples/basics/MultipleWayPointsActivity.kt:
```diff
@@ -0,0 +1,541 @@
+     */
+    private val addedWaypoints = WaypointsSet()
+
+    /**
+     * Debug tool used to play, pause and seek route progress events that can be used to produce mocked location updates along the route.
+     */
+    private val mapboxReplayer = MapboxReplayer()
+
+    /**
+     * Debug tool that mocks location updates with an input from the [mapboxReplayer].
+     */
+    private val replayLocationEngine = ReplayLocationEngine(mapboxReplayer)
```
Just leaving an option here. Both ways work, but you may prefer this new way.

you can delete these and call "MapboxNavigation.startReplayTripSession()"
Access the mapboxReplayer through `mapboxNavigation.mapboxReplayer`

We haven't added it to many examples yet, but you can see it here https://github.com/mapbox/mapbox-navigation-android/blob/main/examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt

VysotskiVadim replied:
thanks, removed. less code is better 🙂

abhishek1508 commented:
@VysotskiVadim This is awesome work 🎉 
There are a few things I would like to point out
- Based on history and user feedback, we want our examples to be as short as possible and to the point.
- We don't need to add any extra features in the example that is not relevant to the example. In this case please remove the any code related to voice, as the examples intention is to showcase multiwaypoints in a route.
- I don't think we need to demonstrate turn by turn experience again. In this case there is no need to show how to replay routes, start navigation, show maneuvers etc  as there are examples showing how to do them. I see this as an example that shows how to add multi waypoints to a route and create different kinds of them.
- We have found customers looking into our examples find it helpful if we add relevant comments. Let's add comments to code snippets that are relevant to this example. This includes the details on what this example does on top of the class name as done in all other examples.
- We should not add any `Utils` to the examples as these examples are directly referred in the public documentation. Adding any other class references will result in compile time issues for the examples and any customers doing a copy paste of the example itself because now they have to resolve `Utils`. The example should be self sufficient. Hence let's remove `Utils.kt` and revert all other examples using this.
- We have agreed that no example would handle `onBackPressed` events. I would encourage to remove it. Any clicks on device back button would result in going back to `MainActivity`

VysotskiVadim commented:
@abhishek1508 , thank you for the feedback!

> I don't think we need to demonstrate turn by turn experience again. In this case there is no need to show how to replay routes, start navigation, show maneuvers, play voice instructions etc as there are examples showing how to do them. The highlight of the example is WaypointsSet and addWaypoint(). We should only focus on these and I encourage you to remove all other irrelevant snippets.

Agree, I didn't do it at the beginning. But then I found out that the maneuver view and the voice instructions are the only components which care about named waypoints. If I get rid of them there will be no difference between regular and named waypoints.

VysotskiVadim commented:
> Silent waypoints do not become actually **silent**
> 
>  device-2021-11-09-144852.mp4

@RingerJK , fixed

abhishek1508 commented:
> But then I found out that the maneuver view and the voice instructions are the only components which care about named waypoints. If I get rid of them there will be no difference between regular and named waypoints.

Agreed but then, this example is doing a great job in demonstrating how to create the regular and named waypoints, populate `.waypointNamesList(addedWaypoints.waypointsNames())`. This inner class `WaypointSet` is the value add as it is showing how to create `waypointIndicesList` and `waypointsNameList`. I would still encourage to remove maneuver and voice and use comments like below:

```
when (val index = binding.waypointTypeView.selectedTabPosition) {
    0 -> addedWaypoints.addRegular(destination)
    // When you add named waypoints, the string you use here inside "" would be shown in `Maneuver` and played in `Voice` instructions.
    1 -> addedWaypoints.addNamed(destination, "Example of the named waypoint")
    // When you add silent waypoints, make sure it is followed by a regular or named waypoint, otherwise silent waypoint is treated as a regular waypoint
    2 -> addedWaypoints.addSilent(destination)
    else -> error("Not supported type of waypoint. index $index")
}
```
We have found that comments like these have been very helpful to the developers looking at the examples while removing the unnecessary noise. Also, things like adding named waypoints would should the way name in maneuver view and play the way name in voice instructions should be a part of public docs. 

cc @VysotskiVadim

VysotskiVadim commented:
@abhishek1508, I see your point. I agree with you. When I'm looking at the example I find it too complex as well. It demonstrates more than just adding multiple waypoints and can be confusing.
I think it's enough to just to say in the comments that a name will be demonstrated in manoeuvre view and played in voice instructions.

I have only one doubt - **testing**. How could we know that the named waypoints works if we don't see nor hear it? I worry that we could accidentally break the example. We won't be able to test named waypoints if I remove all components that display the name.

What do you think about testing? cc @mapbox/navigation-android

abhishek1508 commented:
> I have only one doubt - testing. How could we know that the named waypoints works if we don't see nor hear it? I worry that we could accidentally break the example. We won't be able to test named waypoints if I remove all components that display the name.

This is a fair point. I would recommend the following:

ManeuverApi exposes a method `fun getManeuvers(route: DirectionsRoute, routeLegIndex: Int? = null)`

In the example inside this callback
```
object : RouterCallback {
    override fun onRoutesReady(
        routes: List<DirectionsRoute>,
        routerOrigin: RouterOrigin
    ) {
        setRoute(routes)
        // You can invoke
        val maneuvers = maneuverApi.getManeuvers(route[0])
        Log.d(TAG, "maneuvers: $maneuvers")
   }
```
Instead of adding the entire maneuver view you can just request the maneuvers every time a route is set and log it. I would also recommend to add a comment as to why are you using maneuver API here to log the named waypoint. This way we can remove the entire code related to maneuverview and still be able to test `named waypoint`. This would also facilitate removing complete code snippet on voice and hence no need for turn by turn.

cc @VysotskiVadim

VysotskiVadim commented:
@abhishek1508 , thank you, that should work! I will try it.

Zayankovsky commented on app/src/main/res/values/strings.xml:
```diff
@@ -43,4 +43,13 @@
   <string name="description_turn_by_turn">Render a complete turn-by-turn experience using all relevant Navigation SDK APIs and pre-built UI components.</string>
+
+  <!-- Multiple way points -->
+  <string name="title_multiple_way_points">Navigate thought multiple way points</string>
```
```suggestion
  <string name="title_multiple_way_points">Navigate through multiple way points</string>
```

abhishek1508 commented:
Thanks for addressing the feedback @VysotskiVadim Could you please also add information about what this example does right on top of the class as all other examples do?

VysotskiVadim commented:
> Thanks for addressing the feedback @VysotskiVadim Could you please also add information about what this example does right on top of the class as all other examples do?

Added, thanks! cc @abhishek1508
